core/includes/config.inc

```diff
@@ -839,6 +872,303 @@ interface ConfigStorageInterface {
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function read($name) {
+    if (!$this->exists($name)) {
+      return FALSE;
+    }
+    $data = db_select($this->table, 'c', array('target' => $this->database))
+      ->fields('c', array('data'))
+      ->condition('c.name', $name)
+      ->execute()
+      ->fetchField();
+    try {
+      $data = $this->decode($data);
+      // Remove the config name from the read configuration.
+      if (isset($data['_config_name'])) {
+        unset($data['_config_name']);
+      }
+    }
+    // If an error occurs, catch and rethrow with the file name in the message.
+    catch (ConfigStorageException $e) {
+      throw new ConfigStorageReadException(format_string("The configuration file \"@filename\" is not properly formatted JSON.\n\nContents:\n<pre>@contents</pre>\n", array('@filename' => $name, '@contents' => $data)));
+    }
+    return $data;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function readMultiple(array $names) {
+    $list = array();
+    foreach ($names as $name) {
+      if ($data = $this->read($name)) {
+        $list[$name] = $data;
+      }
+    }
+    return $list;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function write($name, array $data) {
+    // Ensure that the config name is included in the written file.
+    $data = array_merge(array('_config_name' => $name), $data);
+    $data = $this->encode($data) . "\n";
+    $file_path = $this->getFilePath($name);
+    try {
+      db_merge($this->table, array('target' => $this->database))
+        ->key(array('name' => $name))
+        ->fields(array(
+          'name' => $name,
+          'data' => $data,
+          'ctime' => time(),
+        ))
+        ->execute();
+    }
+    catch (\Exception $e) {
+      throw new ConfigStorageException('Failed to write configuration file: ' . $this->getFilePath($name), 0, $e);
+    }
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function delete($name) {
+    if (!$this->exists($name)) {
+      return FALSE;
+    }
+    db_delete($this->table, array('target' => $this->database))
+      ->condition('name', $new_name)
+      ->execute();
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function rename($name, $new_name) {
+    try {
+      db_delete($this->table, array('target' => $this->database))
+        ->condition('name', $new_name)
+        ->execute();
+      db_update($this->table, 'c', array('target' => $this->database))
+        ->fields('c.name', $new_name)
+        ->condition('c.name', $name)
+        ->execute();
+    }
+    catch (\Exception $e) {
+      throw new ConfigStorageException('Failed to rename configuration file from: ' . $this->getFilePath($name) . ' to: ' . $this->getFilePath($new_name), 0, $e);
+    }
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getModifiedTime($name) {
+    $data = db_select($this->table, 'c', array('target' => $this->database))
+      ->fields('c', 'ctime')
```

Comment:
Ow ow ow. `getModifiedTime()` returns the *created time*?? Not your fault, just... dang. A weirdness we picked up in our port of CMI it seems.